### PR TITLE
Run LLVM coverage instrumentation passes before optimization passes

### LINKED ISSUE #83429: Consider moving instrument-coverage-specific behavior in rustc_typeck to backend code
See review comment https://github.com/rust-lang/rust/pull/83307#discussion_r600011365 suggesting we may want to move the `-Z instrument-coverage`-specific behavior from `rustc_typeck` to a place more relevant to LLVM and/or codegen.

`rustc_typeck` appears to be the most central place to apply this fix. Is there a better place, or if not, can we add one?

### PR DESCRIPTION
**Pull request by Amanieu**
This matches the behavior of Clang and allows us to remove several
hacks which were needed to ensure functions weren't optimized away
before reaching the instrumentation pass.

Fixes #83429

cc @richkadel 

r? @tmandry

### CODE REVIEW

File: src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs [hunk 1]
```diff
@@ -28,12 +28,6 @@ pub fn used_inline_function() {
     }
     use_this_lib_crate();
 }
```
Comment (richkadel) [inline src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs:30]: Any theories why this is no longer generating the `Unexecuted instantiation`? It probably doesn't have to be documented here anymore (the absence of the unexpected should be expected, so this is good), but I'm curious why it changed.

Here's a trick for keeping the diffs clean in the `expected` results, above. The line numbers in `llvm-cov show` mess things up, so I avoid changing line numbers whenever possible: Just replace these 6 deleted lines with 6 blank lines.

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 2]
```diff
@@ -548,6 +548,11 @@ pub(crate) unsafe fn optimize(
                     llvm::LLVMRustAddPass(fpm, find_pass("lint").unwrap());
                     continue;
                 }
+                if pass_name == "insert-gcov-profiling" || pass_name == "instrprof" {
+                    // Instrumentation should be inserted before optimization.
```
Comment (richkadel) [inline compiler/rustc_codegen_llvm/src/back/write.rs:552]: Can we say "must" instead of "should" and reference the Issue number?

**Review by richkadel — CHANGES_REQUESTED** [review]
Looks great! Thanks!

File: src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs (hunk 1 above)
Comment (Amanieu) [reply src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs:30]: The `Unexecuted instantiation` re-appears if I undo the changes in `compiler/rustc_typeck/src/collect.rs`. I'm not exactly sure *how* that causes it to re-appear though.

File: compiler/rustc_codegen_llvm/src/back/write.rs (hunk 2 above)
Comment (Amanieu) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: I couldn't find a suitable issue to point to so I just pointed at the Clang code.

File: compiler/rustc_typeck/src/collect.rs [hunk 3]
```diff
@@ -2889,17 +2889,7 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
...
-                        // marked with `#[inline(always)]`, which can break coverage reporting if
-                        // that function was referenced from a coverage map.
-                        //
-                        // FIXME(#83429): Is there a better place, e.g., in codegen, to check and
-                        // convert `Always` to `Hint`?
-                        InlineAttr::Hint
-                    } else {
-                        InlineAttr::Always
-                    }
+                    InlineAttr::Always
```
Comment (richkadel) [inline compiler/rustc_typeck/src/collect.rs:2892]: Oh, before you make this change, we need to test that this doesn't break the brotli_decompressor

See #82875

I was not able to create a coverage test that failed in the way it fails in this issue, but by not forcing inline always, the issue was resolved.

**Review by richkadel — CHANGES_REQUESTED** [review]
Sorry, I just realized that the existing tests in the rust compiler don't confirm that your fix for the inline always actually works. See my comment about how to test this. Thanks!

File: compiler/rustc_codegen_llvm/src/back/write.rs (hunk 2 above)
Comment (richkadel) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: You're updated comment works for me. Thanks!

File: src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs (hunk 1 above)
Comment (richkadel) [reply src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs:30]: Yeah, I assumed it's related somehow.

Thanks for inserting the blank lines. Makes the expected result diffs a lot easier to read!

File: compiler/rustc_typeck/src/collect.rs (hunk 3 above)
Comment (Amanieu) [reply compiler/rustc_typeck/src/collect.rs:2892]: I just tested the example from that issue. It fails on my main rustc (2021-03-24 nightly) but succeeds on my new branch. So the fix for inline always works.

File: src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs (hunk 1 above)
Comment (Amanieu) [reply src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs:30]: Do you want me to remove the blank lines before merging?

Comment (richkadel) [reply src/test/run-make-fulldeps/coverage/lib/used_inline_crate.rs:30]: Not necessary. Thanks.

File: compiler/rustc_typeck/src/collect.rs (hunk 3 above)
Comment (richkadel) [reply compiler/rustc_typeck/src/collect.rs:2892]: That's great. 

Before merging, I'd like to try your patch on some creates I had issues with.

Comment (richkadel) [reply compiler/rustc_typeck/src/collect.rs:2892]: I cherry-picked your branch commits and am testing them now.

**Comment by richkadel** [comment]
I'm having an issue compiling the brotli_decompressor test with anything higher than `-Copt-level=1`. The compiler is much slower when compiling with `-Copt-level=2`. (Level 3 is not much slower than 2 though.)

I don't remember ever having to wait very long for the compiler to compile and link one small program like this.

Is it possible your change could significantly slow down the optimization passes?

```rust
extern crate brotli_decompressor;
use std::io;

fn main() {
    match brotli_decompressor::BrotliDecompress(&mut io::stdin(), &mut io::stdout()) {
        Ok(_) => {}
        Err(e) => println!("Error {:?}", e),
    }
}
```

```shell
$ time ~/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Zinstrument-coverage -Copt-level=1 brotli_covtest.rs -L ~/rust-brotli-decompressor/target/debug/ -L ~/rust-brotli-decompressor/target/debug/deps --edition=2018

real    0m1.111s
user    0m3.136s
sys     0m0.179s

$ time ~/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Zinstrument-coverage -Copt-level=2 brotli_covtest.rs -L ~/rust-brotli-decompressor/target/debug/ -L ~/rust-brotli-decompressor/target/debug/deps --edition=2018

real    0m23.503s
user    0m36.125s
sys     0m0.181s

$ time ~/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Zinstrument-coverage -Copt-level=3 brotli_covtest.rs -L ~/rust-brotli-decompressor/target/debug/ -L ~/rust-brotli-decompressor/target/debug/deps --edition=2018

real    0m23.822s
user    0m36.961s
sys     0m0.211s
```

**Comment by Amanieu** [comment]
I can't reproduce this:
```
time rustc +stage1 -Zinstrument-coverage -Copt-level=3 brotli_covtest.rs -L target/debug/ -L target/debug/deps --edition=2018                                                                                                                                                                                    
________________________________________________________
Executed in    1.45 secs    fish           external
   usr time    2.26 secs  396.00 micros    2.26 secs
   sys time    0.07 secs   46.00 micros    0.07 secs
```

Can you try using `-Ztime-passes -Ztime-llvm-passes` to see what is taking up so much time?

**Comment by richkadel** [comment]
yes. Also, Let me rebuild the compiler after a clean. I want to make sure I don't have some weird residual code influencing this somehow.

**Comment by richkadel** [comment]
I'm not seeing an issue after a clean build. Not sure what happened, but I think it was something buggy in my build. Good to go from my perspective.

@tmandry - Thumbs up from me. 

Thanks Amanieu!

**Comment by tmandry** [comment]
Oh this is great, a big mystery is now solved. Thanks for this!

@bors r+

**Comment by klensy** [comment]
https://perf.rust-lang.org/compare.html?start=65b44b0320e88f5a0608126c36b02be1b840700f&end=6ff482bde5d22a3a0171edb3245327f43cf9b593&stat=max-rss

Big deviations on kessak-debug and unicode_normalization-debug with incr-full, but maybe this is random thing.

**Comment by Amanieu** [comment]
The changes in PR only apply if `-Zinstrument-coverage` is used, so this is very surprising.

**Comment by klensy** [comment]
Currently rust-perf site works slow for me, maybe that affects backend performance collection process too?

**Comment by klensy** [comment]
Checked next pr, the same thing https://perf.rust-lang.org/compare.html?start=2a32abbcdea97c6bf1d0445bc657f16c50ca103b&end=a5029ac0ab372aec515db2e718da6d7787f3d122&stat=max-rss

Maybe some issues with performance collecting, or i checking wrong commits :-(

**Pull request merged**